Libraries/LibJS/Runtime/GeneratorFunction.cpp

```diff
@@ -0,0 +1,129 @@
+
+namespace JS {
+
+GeneratorFunction* GeneratorFunction::create(GlobalObject& global_object, const FlyString& name, const Statement& body, Vector<FlyString> parameters, LexicalEnvironment* parent_environment, LexicalEnvironment* own_environment)
+{
+    return global_object.heap().allocate<GeneratorFunction>(name, body, move(parameters), parent_environment, *global_object.function_prototype(), own_environment);
+}
+
+GeneratorFunction::GeneratorFunction(const FlyString& name, const Statement& body, Vector<FlyString> parameters, LexicalEnvironment* parent_environment, Object& prototype, LexicalEnvironment* own_environment)
+    : ScriptFunction(name, move(body), move(parameters), parent_environment, prototype)
+    , m_own_environment(own_environment)
+{
+}
+
+GeneratorFunction::~GeneratorFunction()
+{
+}
+
+void GeneratorFunction::visit_children(Visitor& visitor)
+{
+    ScriptFunction::visit_children(visitor);
+    if (m_own_environment)
+        visitor.visit(m_own_environment);
+}
+
+LexicalEnvironment* GeneratorFunction::create_environment()
+{
+    if (m_own_environment)
+        return m_own_environment;
+
+    HashMap<FlyString, Variable> variables;
+    for (auto& parameter : m_parameters) {
+        variables.set(parameter, { js_undefined(), DeclarationKind::Var });
+    }
+
+    if (body().is_scope_node()) {
+        for (auto& declaration : static_cast<const ScopeNode&>(body()).variables()) {
+            for (auto& declarator : declaration.declarations()) {
+                variables.set(declarator.id().string(), { js_undefined(), DeclarationKind::Var });
+            }
+        }
+    }
+    if (variables.is_empty())
+        return m_own_environment = m_parent_environment;
+    return m_own_environment = heap().allocate<LexicalEnvironment>(move(variables), m_parent_environment);
+}
+
+Value GeneratorFunction::call(Interpreter& interpreter)
+{
+    auto& argument_values = interpreter.call_frame().arguments;
+    ArgumentVector arguments;
+    for (size_t i = 0; i < m_parameters.size(); ++i) {
+        auto name = parameters()[i];
+        auto value = js_undefined();
+        if (i < argument_values.size())
+            value = argument_values[i];
+        arguments.append({ name, value });
+        interpreter.current_environment()->set(name, { value, DeclarationKind::Var });
+    }
+
+    auto new_body = adopt(*new BlockStatement);
+    auto* block = static_cast<ScopeNode*>(new_body);
+
+    auto& my_block = static_cast<const ScopeNode&>(body());
+
+    for (auto& node : my_block.children()) {
+        block->append(node);
+    }
+    auto iterator = heap().allocate<Iterator>(heap().allocate<GeneratorFunction>(name(), *new_body, parameters(), m_parent_environment, *prototype(), create_environment()), [&interpreter, argument_values, arguments](Object* generator_object, const Vector<Value>& next_arguments) -> Iterator::IteratorResult {
+        auto* gen = static_cast<GeneratorFunction*>(generator_object);
```

Comment:
Please use full words like `generator`.
Also, how do we know it's safe to `static_cast` to a `GeneratorFunction*` here?

Reply:
The `Iterator` gives you back whatever you pass as the first argument.